[안재민] Sprint9

mini-1018 opened the pull request:
## 요구사항

### 기본

### 공통

- [x] Github에 위클리 미션 PR을 만들어 주세요.
- [x] React 혹은 Next.js를 사용해 진행합니다.
- [x] 프론트엔드에서 api 요청 구현은 TanStack React Query를 활용해 주세요.
- [x] API는 https://panda-market-api.vercel.app/를 사용합니다. 명세는 https://panda-market-api.vercel.app/docs를 확인해 주세요.

### 로그인/회원가입 페이지

- [x] JavaScript로 구현한 로그인/회원가입 페이지를 React.js 혹은 Next.js로 마이그레이션해 주세요.
### 로그인 페이지

- [x] "회원 가입하기"를 클릭하면 회원가입 페이지로 이동해 주세요.
- [x] 로그인 실패하는 경우, 이메일 input 아래에 "이메일을 확인해 주세요.", 비밀번호 input 아래에 "비밀번호를 확인해 주세요." 에러 메시지를 표시해 주세요.
- [x] 로그인 버튼이 활성화된 후, 로그인 버튼 클릭 또는 Enter키 입력으로 로그인 실행합니다.
- [x] "/api/signIn"으로 POST 요청해서 성공 응답을 받으면 중고 마켓 페이지로 이동합니다. 참고로 JWT로 구현되어 있습니다.
- [ ] 실패할 경우, 실패 메시지를 모달을 통해 표시합니다.

### 회원가입 페이지

- [x] "회원 가입하기"를 클릭하면 '/signin' 페이지로 이동합니다
- [x] 회원가입 버튼 클릭 또는 Enter키 입력으로 회원가입을 실행합니다.
- [x] 비밀번호 input과 비밀번호 확인 input의 값이 다른 경우, 비밀번호 확인 input 아래에 "비밀번호가 일치하지 않아요." 에러 메시지를 표시해 주세요.
- [x] 버튼이 활성화된 후, 회원가입은 "/api/signUp" POST 요청해서 진행합니다. 참고로 JWT로 구현되어 있습니다.
- [x] 회원가입 성공 응답을 받으면 중고마켓 페이지로 이동합니다.
- [ ] 실패할 경우, 실패 메시지를 모달을 통해 표시합니다.
로그인, 회원가입 페이지 공통

- [x] 눈 모양 아이콘 클릭 시 비밀번호의 문자열이 보이기도 하고, 가려집니다.
- [x] 비밀번호의 문자열이 가려질 때는 눈 모양 아이콘에는 사선이 그어져 있고, 비밀번호의 문자열이 보일 때는 사선이 없는 눈 모양 아이콘이 보입니다.
- [x] 소셜 로그인에 구글 아이콘 클릭 시 'https://www.google.com/', 카카오 아이콘 클릭 시 '[https://www.kakaocorp.com/page'](https://www.kakaocorp.com/page%E2%80%99)로 이동합니다.
- [x] 로그인/회원가입 시 성공 응답으로 받은 accessToken을 로컬 스토리지에 저장합니다.
- [x] 로그인/회원가입 페이지에 접근 시 로컬 스토리지에 accessToken이 있는 경우 '/folder' 페이지로 이동합니다.

### GNB

- [x] 상단 내비게이션 바에 프로필 영역은 인가된 경우, 유저 정보 API를 활용해 주세요.
- [x] 인가되지 않았을 경우 "로그인" 버튼이 보이게 해 주세요.

### 상품 상세 페이지

- [x] PC, Tablet, Mobile 디자인에 해당하는 상품 상세 페이지를 만들어 주세요.
- [x] 상품 상세 페이지 url path는 "/items/{itemId)"로 설정하세요.
- [x] '목록으로 돌아가기' 버튼 클릭 시 중고마켓 페이지 "/items"로 이동합니다.
- [x] 상품 상세 데이터는 '/products/{productId}' GET 메서드 사용해 불러오세요. 이때, 상품 상세 조회는 인가된 사용자만 이용할 수 있도록 합니다.
- [ ] 상품에 대한 댓글 조회도 가능합니다.
- [ ] 상품 수정 및 삭제 기능을 API를 활용해 구현합니다. 이때, 인가된 사용자만 이용할 수 있도록 합니다.
- [ ] 상품 수정은 '/products/{productId}' PUT을 사용합니다.
- [ ] 상품 삭제는 '/products/{productId}' DELETE를 사용합니다.
- [ ] 상품 삭제 전, 확인 모달을 띄워주세요.
- [ ] 상품에 대한 좋아요 및 좋아요 취소 기능을 https://panda-market-api.vercel.app/docs에 명세된 '/products/{productId}/favorite' POST & DELETE 활용해 구현합니다. 이때 인가된 사용자만 좋아요 기능을 이용할 수 있도록 합니다.
- [ ] 댓글 생성 및 삭제 기능을 API를 활용해 구현합니다. 이때, 인가된 사용자만 이용할 수 있도록 합니다.
- [ ] 댓글 수정은 https://panda-market-api.vercel.app/docs에 명세된 '/products/{productId}/comments' PUT을 사용합니다.
- [ ] 댓글 삭제는 https://panda-market-api.vercel.app/docs에 명세된 '/products/{productId}/comments' DELETE를 사용합니다.
### 심화

### 로그인 및 회원가입 페이지 공통

- [x] 로그인, 회원가입 기능에 [react-hook-form](https://react-hook-form.com/)을 활용해 주세요.
- [ ] 브라우저에 현재 보이는 화면의 영역(viewport) 너비를 기준으로 분기되는 반응형 디자인을 적용합니다.
PC: 1200px 이상
Tablet: 744px 이상 ~ 1199px 이하
Mobile: 375px 이상 ~ 743px 이하
375px 미만 사이즈의 디자인은 고려하지 않습니다
유저 기능

- [ ] 리퀘스트 헤더에 인증 토큰을 첨부할 때 [axios interceptors](https://axios-http.com/kr/docs/interceptors)를 활용해 주세요. (axios를 사용하지 않는다면 이와 유사한 기능을 활용해 주세요.)
React-Query로 마이그레이션

- [x] fetch 혹은 axios로 구현된 기존의 API 요청 코드를 React-Qeury로 마이그레이션 합니다.
로딩 및 에러 핸들링

- [ ] 로딩 인디케이터와 에러 메시지를 구현합니다.
- [ ] 상품 목록 및 상품 상세 데이터를 Prefetching 합니다

### 상품 데이터 캐싱 및 업데이트

- [ ] React Query의 캐싱 기능을 활용하여 데이터 로딩 시간을 최소화합니다.
- [ ] 상품 목록 페이지에서 데이터의 실시간 업데이트를 위해 적절한 Query Refresh 설정을 적용합니다.

## 주요 변경사항

- 현재 로그인, 회원가입 페이지 완료하였고 기능으로는 로그인, 회원가입 시 토큰 정상 발급 확인하였습니다.
- 저번 미션 때와 동일하게 zustand, react query 사용하였고 로그인, 회원가입에는 리액트 훅 폼 적용하였습니다.

## 스크린샷

![image](이미지url)

## 멘토에게

- 쿼리 클라이언트 동작 방식과 리액트 훅 폼(watch가 동작하지 않는 현상)에 대해 멘토링 시간에 질문 드리겠습니다.
- comment 컴포넌트 관련해서 기존에는 article폴더에 ArticleComment 이런식으로 작성했지만 sprint8 게시글 댓글 / sprint9 상품 문의 부분의 디자인이 같아 재사용성을 위해 수정 하였습니다. 특히 입력인풋,등록버튼 / 리스트를 나눠 불필요한 렌더링을 줄여보았습니다. 또한 개인적으로 하위 컴포넌트에 옵션에 따른 분기로직을 자꾸 만들고 싶지않아 리액트 쿼리부분에 따라 CommentQuery.js 처럼 나눠봤습니다. 저번에 찝어주셨던 부분중에 컴포넌트는 export default 하는게 좋다고 하셔서 제가 구상한 방법은 통상 쓰이는 방법이 아닌지 궁금합니다.
- 전역관리로 인해 props 드릴링 걱정이 없어 버튼 컴포넌트를 ActionButton 하나만 남기려고 합니다. ActionButton의 온클릭 이벤트를 전역으로 하나 관리해볼까 하는데 괜찮은 방법인지 궁금합니다. 
- 저번에 지적해주셨던 것들(버튼, 쿼리 키 등)아직 전부 깔끔하게 수정하지 못하였습니다. 중간중간 수정하도록하겠습니다.
- 화, 수에도 진행한 만큼 업데이트 하도록 하겠습니다. 항상 감사드립니다.

mini-1018 committed:
feat : 라우터 그룹

mini-1018 committed:
feat : 로그인, 회원가입 api

mini-1018 committed:
feat : 이미지, 레이아웃 수정

mini-1018 committed:
feat : 폼 인풋 / 유효성, 뮤테이션 훅

mini-1018 committed:
feat : 로그인, 회원가입 폼

mini-1018 committed:
feat : 댓글, 문의 입력 관련 comment 컴포넌트 수정(추가)

mini-1018 committed:
feat : 토큰 설정, 로그인 상태관리 추가

wildCodingWarrior commented on app/(nav-layout)/items/[itemId]/page.jsx:
```diff
@@ -0,0 +1,82 @@
+            <div className={styles['product-info']}>
+              <div className={styles['product-meta']}>
+                <div className={styles['product-user-image']}>
+                  {product.ownerNickname.image ? (
+                    <Image
+                      src={product.ownerNickname.image}
+                      fill
+                      alt="user-image"
+                    />
+                  ) : null}
+                </div>
```
P2
삼항연산자 사용하면서 false 조건 값일 때가 null이면 그냥 &&로 연결해서 사용하는게 가독성이 더 괜찮을 것 같아요.

mini-1018 replied:
감사합니다. 수정완료입니다!

wildCodingWarrior commented on src/hooks/comment/useArticleComments.js:
```diff
@@ -0,0 +1,19 @@
+import { useQuery } from '@tanstack/react-query';
+import { getComment } from '@utils/api/api';
+
+export const useArticleComments = ({ id, initialData }) => {
+  const {
+    data: articleComments,
+    isLoading,
+    error,
+  } = useQuery({
+    queryKey: ['articleComments', id],
+    queryFn: async () => getComment(id),
```
queryKey 따로 Constant/queryKeys에서 관리하고 있는거 가져와서 쓰면 더 좋겠네요

mini-1018 replied:
queryKey: [QUERY_KEYS.ARTICLE_COMMENTS, id], 이런식으로 변경 완료입니다!

wildCodingWarrior commented on src/utils/api/codeit.js:
```diff
@@ -0,0 +1,52 @@
+
+    if (refreshToken) {
+      localStorage.setItem('refreshToken', refreshToken);
+    }
+
+    return response.data;
+  } catch (error) {
+    console.error(error);
+  }
+}
+
+export async function postSignup(params) {
+  try {
+    const response = await codeitApi.post('/auth/signUp', params);
+
+    return response.data;
+  } catch (error) {
+    console.error(error);
+  }
+}
+
+export async function getProducts(id) {
+  try {
+    const response = await codeitApi.get(`/products/${id}`);
+    return response.data;
+  } catch (error) {
+    console.error(error.status);
+  }
+}
+
+export async function getProductList(query) {
+  try {
+    const response = await codeitApi.get(`/products`, {
+      params: query,
+    });
+    return response.data.list;
+  } catch (error) {
+    console.error(error.status);
+  }
+}
```
P4
혹시 NextAuth를 사용해보실 생각이 있으신가요?

사실 useQuery로 관리하는 것 보다 직접적으로 이렇게 로그인이나 회원가입이 있을 경우 NextAuth를 사용하는게 일반적이기는 합니다. 주강사님한테 한번만 여쭤보시고 적용할만하다고 말씀하시면 시도해보세요 (안해도 됨)

wildCodingWarrior reviewed (changes requested):
멘토링 시간에 구두로 리뷰를 한 부분이 많네요!

전체적으로 코드가 깔끔해서 리뷰할 부분이 많이 없었습니다만, 코드리뷰 진행해두었습니다.

mini-1018 committed:
feat : 인증, 토큰 api 추가

mini-1018 committed:
feat : 미들웨어 추가

미션 요구사항에 따라 /items/{itemId}는 인가된(토큰이 있는지 없는지만 구현) 사용자만 보여질 수 있도록 미들웨어 설정

mini-1018 committed:
feat : zustand 인증 스토어 추가

모든 경로의 페이지 로드 시(루트 레이아웃에 적용) 쿠키에 저장된 토큰으로 로그인유지 또는 토큰이 없거나 유효하지 않을 시 재요청, 로그인페이지로 리다이렉트

mini-1018 commented:
## 요구사항

### 기본

### 공통

- [x] Github에 위클리 미션 PR을 만들어 주세요.
- [x] React 혹은 Next.js를 사용해 진행합니다.
- [x] 프론트엔드에서 api 요청 구현은 TanStack React Query를 활용해 주세요.
- [x] API는 https://panda-market-api.vercel.app/를 사용합니다. 명세는 https://panda-market-api.vercel.app/docs를 확인해 주세요.

### 로그인/회원가입 페이지

- [x] JavaScript로 구현한 로그인/회원가입 페이지를 React.js 혹은 Next.js로 마이그레이션해 주세요.
### 로그인 페이지

- [x] "회원 가입하기"를 클릭하면 회원가입 페이지로 이동해 주세요.
- [x] 로그인 실패하는 경우, 이메일 input 아래에 "이메일을 확인해 주세요.", 비밀번호 input 아래에 "비밀번호를 확인해 주세요." 에러 메시지를 표시해 주세요.
- [x] 로그인 버튼이 활성화된 후, 로그인 버튼 클릭 또는 Enter키 입력으로 로그인 실행합니다.
- [x] "/api/signIn"으로 POST 요청해서 성공 응답을 받으면 중고 마켓 페이지로 이동합니다. 참고로 JWT로 구현되어 있습니다.
- [ ] 실패할 경우, 실패 메시지를 모달을 통해 표시합니다.

### 회원가입 페이지

- [x] "회원 가입하기"를 클릭하면 '/signin' 페이지로 이동합니다
- [x] 회원가입 버튼 클릭 또는 Enter키 입력으로 회원가입을 실행합니다.
- [x] 비밀번호 input과 비밀번호 확인 input의 값이 다른 경우, 비밀번호 확인 input 아래에 "비밀번호가 일치하지 않아요." 에러 메시지를 표시해 주세요.
- [x] 버튼이 활성화된 후, 회원가입은 "/api/signUp" POST 요청해서 진행합니다. 참고로 JWT로 구현되어 있습니다.
- [x] 회원가입 성공 응답을 받으면 중고마켓 페이지로 이동합니다.
- [ ] 실패할 경우, 실패 메시지를 모달을 통해 표시합니다.
로그인, 회원가입 페이지 공통

- [x] 눈 모양 아이콘 클릭 시 비밀번호의 문자열이 보이기도 하고, 가려집니다.
- [x] 비밀번호의 문자열이 가려질 때는 눈 모양 아이콘에는 사선이 그어져 있고, 비밀번호의 문자열이 보일 때는 사선이 없는 눈 모양 아이콘이 보입니다.
- [x] 소셜 로그인에 구글 아이콘 클릭 시 'https://www.google.com/', 카카오 아이콘 클릭 시 '[https://www.kakaocorp.com/page'](https://www.kakaocorp.com/page%E2%80%99)로 이동합니다.
- [x] 로그인/회원가입 시 성공 응답으로 받은 accessToken을 로컬 스토리지에 저장합니다.
- [x] 로그인/회원가입 페이지에 접근 시 로컬 스토리지에 accessToken이 있는 경우 '/folder' 페이지로 이동합니다.

### GNB

- [x] 상단 내비게이션 바에 프로필 영역은 인가된 경우, 유저 정보 API를 활용해 주세요.
- [x] 인가되지 않았을 경우 "로그인" 버튼이 보이게 해 주세요.

### 상품 상세 페이지

- [x] PC, Tablet, Mobile 디자인에 해당하는 상품 상세 페이지를 만들어 주세요.
- [x] 상품 상세 페이지 url path는 "/items/{itemId)"로 설정하세요.
- [x] '목록으로 돌아가기' 버튼 클릭 시 중고마켓 페이지 "/items"로 이동합니다.
- [x] 상품 상세 데이터는 '/products/{productId}' GET 메서드 사용해 불러오세요. 이때, 상품 상세 조회는 인가된 사용자만 이용할 수 있도록 합니다.
- [ ] 상품에 대한 댓글 조회도 가능합니다.
- [ ] 상품 수정 및 삭제 기능을 API를 활용해 구현합니다. 이때, 인가된 사용자만 이용할 수 있도록 합니다.
- [ ] 상품 수정은 '/products/{productId}' PUT을 사용합니다.
- [ ] 상품 삭제는 '/products/{productId}' DELETE를 사용합니다.
- [ ] 상품 삭제 전, 확인 모달을 띄워주세요.
- [ ] 상품에 대한 좋아요 및 좋아요 취소 기능을 https://panda-market-api.vercel.app/docs에 명세된 '/products/{productId}/favorite' POST & DELETE 활용해 구현합니다. 이때 인가된 사용자만 좋아요 기능을 이용할 수 있도록 합니다.
- [ ] 댓글 생성 및 삭제 기능을 API를 활용해 구현합니다. 이때, 인가된 사용자만 이용할 수 있도록 합니다.
- [ ] 댓글 수정은 https://panda-market-api.vercel.app/docs에 명세된 '/products/{productId}/comments' PUT을 사용합니다.
- [ ] 댓글 삭제는 https://panda-market-api.vercel.app/docs에 명세된 '/products/{productId}/comments' DELETE를 사용합니다.
### 심화

### 로그인 및 회원가입 페이지 공통

- [x] 로그인, 회원가입 기능에 [react-hook-form](https://react-hook-form.com/)을 활용해 주세요.
- [ ] 브라우저에 현재 보이는 화면의 영역(viewport) 너비를 기준으로 분기되는 반응형 디자인을 적용합니다.
PC: 1200px 이상
Tablet: 744px 이상 ~ 1199px 이하
Mobile: 375px 이상 ~ 743px 이하
375px 미만 사이즈의 디자인은 고려하지 않습니다
유저 기능

- [ ] 리퀘스트 헤더에 인증 토큰을 첨부할 때 [axios interceptors](https://axios-http.com/kr/docs/interceptors)를 활용해 주세요. (axios를 사용하지 않는다면 이와 유사한 기능을 활용해 주세요.)
React-Query로 마이그레이션

- [x] fetch 혹은 axios로 구현된 기존의 API 요청 코드를 React-Qeury로 마이그레이션 합니다.
로딩 및 에러 핸들링

- [ ] 로딩 인디케이터와 에러 메시지를 구현합니다.
- [ ] 상품 목록 및 상품 상세 데이터를 Prefetching 합니다

### 상품 데이터 캐싱 및 업데이트

- [ ] React Query의 캐싱 기능을 활용하여 데이터 로딩 시간을 최소화합니다.
- [ ] 상품 목록 페이지에서 데이터의 실시간 업데이트를 위해 적절한 Query Refresh 설정을 적용합니다.

## 주요 변경사항

- 로그인유지는 zustand 이용하여 루트 레이아웃에 useEffect로 모든 url로 접근 시 쿠키에 저장된 토큰을 이용하여 처음 한번 실행되도록 하였습니다. (쿠키에 저장된 토큰 -> authorize가 필요한 명세인 /users/me에 요청을 보내고 user 정보를 받아오도록 구성, 토큰 만료 시 리프레시 토큰으로 재요청, 전부 불가할 시 로그인 페이지로 이동)
- 요구사항 중 상품상세 페이지와 그 안의 수정 삭제 기능들 모두 인가된 사용자만 이라는 조건이 있어 이 부분은 next.js 공식문서 미들웨어 활용해보았습니다. /items/:path*로 단순히 토큰이 있을 시에만 보여지고 없을 시 로그인페이지로 리다이렉트 하도록 구성하였습니다.

## 스크린샷

![image](https://github.com/user-attachments/assets/f433829b-422c-4055-b5c5-ea85a10c427c)
상단 네비게이션 오른쪽 로그인 유지 시 로그인 버튼 대신 유저닉네임 출력되도록 설정

## 멘토에게

※유저기능 구현 과정
요구사항에는 로컬스토리지를 활용하는 방식으로 되어있어 로컬스토리지를 활용하던 중 서버에서는 로컬스토리지에 접근할 수 없어 CSR 방식으로 밖에 할 수 없었습니다. 그로 인한 문제점이

1. 클라이언트에서 처리되므로 useEffect가 실행되기 전까지 잠시 깜박이는 현상과 지연발생 시 잠시 페이지가 보이는 현상
2. 지금은 상관없지만 클라이언트측에서 보여주거나 제한하므로 SEO측면에서 문제가 발생할거라 생각했습니다.

그래서 SSR방식으로 알아보게 되었고

1. 로그인 유지는 zustand로 루트 레이아웃에서 토큰활용하여 한번만 실행되도록 useEffect사용
2. 요구사항의 특정 페이지는 인가된 사용자만 볼 수 있도록 미들웨어 활용해보았습니다.

생각보다 유저기능 관련해서 방식과 장단점이 있어 이해도가 많이 필요한 것 같습니다. 배움이 짧아 온전히 이해하지 못하고 구현해서 미흡한점이 너무너무 많습니다.
로컬스토리지에 토큰을 보관하는게 요구사항이였고 로컬스토리지를 이용하면서 SEO를 신경쓰다보니 미들웨어까지 알아보게 되었는데 CSR과 미들웨어의 보안적 단점을 모두 갖는 좋지않은 방법으로 접근한 것 같습니다. 시간을 많이 썼는데 결국 토큰은 쿠키에 저장하고 SSR로 구현하는게 맞다는 생각이 듭니다.. 현업에서 유저기능 구현은 CSR, SSR, 미들웨어 중에 어떤 방식을 일반적으로 사용하는지 궁금합니다. 또한 여기저기 헤메느라 nextAuth 언급해주셨던 걸 깜박했는데 다음 과제에서 nextAuth활용하도록 하겠습니다.

- sprint9은 이번에 배운 유저기능 구현이 핵심이라 생각하였고 아직 로그인 유지 방식, 토큰 활용 등 너무 미숙한점이 많아 상대적으로 많이 구현해본 파트는 우선순위에서 밀려 시간이 부족하여 미구현하였습니다.. (로그인 회원가입 성공 실패 모달 미구현, 댓글 수정삭제 미구현)

wildCodingWarrior commented on app/middleware.js:
```diff
@@ -0,0 +1,22 @@
+import { NextResponse } from 'next/server';
+
+export function middleware(request) {
+  const accessToken = request.cookies.get('accessToken')?.value;
+
+  if (!accessToken) {
+    return NextResponse.redirect(new URL('/login', request.url));
+  }
+
+  const requestHeaders = new Headers(request.headers);
+  requestHeaders.delete('cookie');
+
+  return NextResponse.next({
+    request: {
+      headers: requestHeaders,
+    },
+  });
+}
+
+export const config = {
+  matcher: ['/items/:path*'],
+};
```
NextAuth를 사용하지 않고, 서버를 직접적으로 건들 수 없는 상황에서 쿠키를 읽어서 이것저것 하는 노력한 것 만으로 좋네요.

wildCodingWarrior commented on src/shared/store/auth/auth.js:
```diff
@@ -1,24 +1,90 @@
+import { getUser, postRefreshToken } from '@utils/api/codeit';
 import { create } from 'zustand';
 
+function getCookie(name) {
+  const value = `; ${document.cookie}`;
+  const parts = value.split(`; ${name}=`);
+  if (parts.length === 2) return parts.pop().split(';').shift();
+  return null;
+}
+
+function setCookie(name, value, days) {
+  const expires = new Date(
+    Date.now() + days * 24 * 60 * 60 * 1000
+  ).toUTCString();
+  document.cookie = `${name}=${value}; expires=${expires}; path=/; secure; samesite=strict;`;
+}
+
+function deleteCookie(name) {
+  document.cookie = `${name}=; Max-Age=0; path=/;`;
+}
+
```
Next.js에서는 cookies 활용하기 위한 method를 따로 제공하긴 합니다.
https://nextjs.org/docs/app/api-reference/functions/cookies#deleting-cookies

wildCodingWarrior replied:
다만 이를 클라이언트에서 하는건 무리가 있고, NextAuth처럼 RouteHandler에서 보통 관리를 하죠. 쿠키는 서버에서 다룰 수 있으니까요.

wildCodingWarrior reviewed (approved):
보안적인 측면에서 토큰은 쿠키에 저장하는 것이 더 안전한 것이 맞습니다. 
초기 로딩 성능 및 SEO 고려할 때 SSR 방식이 더 적합한 것도 맞습니다.
Next.js의 미들웨어 활용하는 것도 올바르고 괜찮은 접근 방식이구요.

현업에서는 일반적으로 HTTP-only 쿠키를 사용하는데, 이 말은 오직 서버에서만 읽고 수정 가능한 쿠키를 의미합니다. 생각하시고 접근하신 바가 전부 옳습니다!

인증은 보통 SSR + CSR 혼합해서 사용하는게 맞고, 보통 초기 로드를 SSR을 이용해서 처리하고, 이후 상호작용이 필요한 부분(로그인 / 로그아웃 / 회원 탈퇴 등)에서는 CSR을 이용하죠. (사실 다 NextAuth 쓰면 됩니다.)
